[SGD] v1 to v2 Migration Guide (#18887)
ray-project/ray: python/ray/util/sgd/v2/migration/sgd-migration-guide.md

> # Ray SGD: v1 to v2 migration guide

Reviewer: Let's be sure to link this in the docs in every SGD v1 page?

> In Ray 1.7, we are rolling out a new and more streamlined version of Ray SGD. Ray SGD v2 focuses on usability and composability- it has a much simpler API, has support for more deep learning backends, integrates better with other libraries in the Ray ecosystem, and will continue to be actively developed with more features. 

Reviewer 2: ```suggestion
In Ray 1.7, we are rolling out a new and more streamlined version of Ray SGD. Ray SGD v2 focuses on usability and composability - it has a much simpler API, has support for more deep learning backends, integrates better with other libraries in the Ray ecosystem, and will continue to be actively developed with more features. 
```

Author: Changed!

> This guide will cover to easily migrate existing code from Ray SGD v1 to Ray SGD v2. If you are new to Ray SGD as a whole, you should get started with Ray SGD v2 directly: https://docs.ray.io/en/master/raysgd/v2/raysgd.html. 

Reviewer 2: ```suggestion
This guide will help you easily migrate existing code from Ray SGD v1 to Ray SGD v2. If you are new to Ray SGD as a whole, you should get started with Ray SGD v2 directly: https://docs.ray.io/en/master/raysgd/v2/raysgd.html. 
```

Author: Fixed!

For a full list of features that Ray SGD v2 provides, please check out the user guide: https://docs.ray.io/en/master/raysgd/v2/user_guide.html. 

If there are any issues or anything missing with this guide or any feedback on Ray SGD v2 overall, please file a Github issue on the Ray repo: https://github.com/ray-project/ray/issues!

> ## What are the API differences?

Reviewer 3: Change this section to `Overview`? The transition from this section to the following section wasn't so clear and I think the reason why was that this section doesn't include concrete migration steps while the following sections do.

Author: I added the small snippet at the end saying what the next sections are going to be so that should cover the confusion I believe.

There are 2 primary API differences between Ray SGD v1 and v2.

1. There is a single `Trainer` interface for all backends (torch, tensorflow, horovod), and the backend is simply specified via an argument: `Trainer(backend="torch")`, `Trainer(backend="horovod")`, etc. Any features that we add to Ray SGD will be supported for all backends, and there won't be any API divergence like there was with a separate `TorchTrainer` and `TFTrainer`.

> 2. The `TrainingOperator` and creator functions are replaced by a generic user-defined training function. You no longer have to make your training logic fit into a restrictive interface. In Ray SGD v2, you simply have to provide a training function that describes the full logic for your training execution and this will be distributed by Ray SGD. 

Reviewer 3: ```suggestion
2. The `TrainingOperator` and creator functions are replaced by a more natural user-defined training function. You no longer have to make your training logic fit into a restrictive interface. In Ray SGD v2, you simply have to provide a training function that describes the full logic for your training execution and this will be distributed by Ray SGD. 
```

Author: FIxed!

```python
# Torch Example
def train_func_distributed():
    num_epochs = 3
    model = NeuralNetwork()
    model = DistributedDataParallel(model)
    loss_fn = nn.MSELoss()
    optimizer = optim.SGD(model.parameters(), lr=0.1)

    for epoch in range(num_epochs):
        output = model(input)
        loss = loss_fn(output, labels)
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
        print(f"epoch: {epoch}, loss: {loss.item()}")

from ray.sgd import Trainer

trainer = Trainer(backend="torch", num_workers=4)
trainer.start()
results = trainer.run(train_func_distributed)
trainer.shutdown()
```
Currently, this means that the user is now responsible for modifying their code to support distributed training (specifying `DistributedDataParallel` for `torch` or `MultiWorkerMirroredStrategy` for `tensorflow`) as opposed to having this be automatically handled internally. However, we have plans to provide utilities that you can use to automatically handle these recipes for you.

> 

Reviewer 3: It would be nice to end this section with a quick list of what's upcoming in the doc so the user knows what to expect. I found myself asking "what about reporting or checkpointing" while reading the `Training Logic` section, which intentionally covers just the pure training aspect.

Example:
```
In the following sections, we will guide you through the steps to migrate:
1. Training Logic
2. Interacting with Trainer state (intermediate metrics, checkpointing)
3. Hyperparameter Tuning with Ray Tune
```

Author: Added!

## Training Logic
The main change you will have to make is how you define your training logic. In Ray SGD v1, the API for defining training logic differed for `TorchTrainer` vs. `TFTrainer`, so the steps to migrate will be different for each of these.

### PyTorch
In v1, the training logic is defined through the `train_epoch` and `train_batch` methods of a `TrainingOperator` class which is passed into the `TorchTrainer`. To migrate to Ray SGD v2, there are 2 options: 
1. If you liked having your training logic in the `TrainingOperator`, you can continue to use the `TrainingOperator` with Ray SGD v2.

> 2. If you felt the `TrainingOperator` is too unnecessary and complex, or you had to customize it extensively, you can define your own training function.

Reviewer 3: nit: Should we reverse the order of these two and encourage the user to define their own training function?  Specifically because we plan to deprecate `TrainingOperator` in the future.

Author: Yes good, point, made the change!

#### Alternative 1: Continue to use `TrainingOperator`
You can define a training function that instantiates your `TrainingOperator` and you can call methods directly on the operator object.

So instead of 
```python
class MyTrainingOperator(TrainingOperator):
	...

trainer = TorchTrainer(training_operator_cls=MyTrainingOperator, num_workers=4, use_gpu=True)

num_epochs=10
for _ in range(num_epochs):
	trainer.train()
	trainer.validate()

final_model = trainer.get_model()
```

you would do
```python
class MyTrainingOperator(TrainingOperator):
	...

def train_func(config):
	device = torch.device(f"cuda:{sgd.local_rank()}" if
                  torch.cuda.is_available() else "cpu")
	if torch.cuda.is_available():
		torch.cuda.set_device(device)
	
	operator_cls = config["operator_cls"]
	
	# Set the args to whatever values you want.

> 	training_operator = operator_cls(

Reviewer 3: Can we just directly use `MyTrainingOperator` in here? Doesn't seem like we need to config this.

Author: Good point, fixed!

		config=config, 
		world_rank=sgd.world_rank(), 
		local_rank=sgd.local_rank(), 
		is_distributed=True,
		device=device,
		use_gpu=True,
		wrap_ddp=True,
		add_dist_sampler=True 

	training_operator.setup(config)

	for idx in range(config["num_epochs"]):
		train_loader = training_operator._get_train_loader()
		# If using DistributedSampler, set the epoch here.
		train_loader.set_epoch(idx)
		training_operator.train_epoch(epoch_idx=idx, iter(train_loader))

		validation_loader = training_operator._get_validation_loader()
		training_operator.validate(iterator=iter(validation_loader))
	
	if sgd.world_rank() == 0:
		return training_operator._get_original_models()
	else:

> 		return None

Reviewer 3: nit: Should we use `save_checkpoint` API instead?

Reviewer 3: This is also a note-to-self since it doesn't look like we do a good job describing how to use `return` in the v2 docs.

Author: Hmm I think for this simple case we should recommend just returning the model.

```

#### Alternative 2: Custom Training Function
If you want to no longer use the `TrainingOperator`, you can define your own custom training function, and use only the parts from `TrainingOperator.train_epoch`, `TrainingOperator.setup`, and `TrainingOperator.validate` that are necessary for your application.

You can see a full example on how to port over regular PyTorch DDP code to Ray SGD here: https://docs.ray.io/en/master/raysgd/v2/user_guide.html#update-training-function

### Tensorflow
The API for `TFTrainer` uses creator functions instead of a `TrainingOperator` to define the training logic. To port over Ray SGD v1 Tensorflow code to v2 you can do the following:
```python
def train_func(config):
	train_dataset, val_dataset = data_creator(config)
	strategy = MultiWorkerMirroredStrategy()
	with strategy.scope():
		model = model_creator(config)
	
	for epoch_idx in range(config["num_epochs"]):
		model.fit(train_dataset)
	
	if sgd.world_rank() == 0:
		return model
	else:
		return None

trainer = Trainer(backend="tensorflow", num_workers=4, config={"num_epochs": 3, ...})
trainer.start()
model = trainer.run(train_func)
```

## Interacting with the `Trainer`

> In Ray SGD v1, the user can iteratively call `trainer.train()` or `trainer.validate()` for each epoch, and can then interact with the trainer to get certain state (model, checkpoints, results, etc.). In Ray SGD v2, this is replaced by a single training function that defines the full training & validation loop for all epochs. There are 3 ways to get state during or after the training execution:

Reviewer 2: Overall comment - we should either pick the third person perspective ("the user") or the second person one ("you") and stick with it thought the entire doc.

Reviewer 3: Some of this would be good to move to the previous `Training Function` section:
> In Ray SGD v1, the user can iteratively call `trainer.train()` or `trainer.validate()` for each epoch
> In Ray SGD v2, this is replaced by a single training function that defines the full training & validation loop for all epochs.

Author: Good point, fixed! @matthewdeng I added a point about this in the API differences.
